java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -311,17 +357,34 @@ public PutResult getResult() {
    */
   public interface ClientStreamListener {
 
-    public void putNext();
+    /**
+     * Send the current data in the corresponding {@link VectorSchemaRoot} to the server.
+     */
+    void putNext();
 
-    public void error(Throwable ex);
+    void putNext(byte[] appMetadata);
```

Comment:
what do you think about using a ArrowBuf here instead? No strong opinion, just thinking that would match the rest of the off-heapness of Arrow.

Reply:
That's more consistent with the rest of Arrow, and with the C++ implementation (which uses `arrow::Buffer`, not `uint8_t*`). We'll have to be careful with making it clear who owns the buffer at what point, though.